specs/waku/v2/waku-v2-rpc-api.md

```diff
@@ -0,0 +1,459 @@
+
+#### WakuRelayMessage
+
+`WakuRelayMessage` is an `Object` containing the following fields:
+| Field | Type | Inclusion | Description |
+| ----: | :--: | :--: |----------- |
+| `payload` | `String` | mandatory | The payload being relayed |
+| `contentTopic` | `Number` | optional | Message content topic for optional content-based filtering |
+
+### `post_waku_v2_relay_v1_message`
+
+The `post_waku_v2_relay_v1_message` method publishes a message to be relayed on a [PubSub `topic`](https://github.com/libp2p/specs/blob/master/pubsub/README.md#the-topic-descriptor)
+
+#### Parameters
+
+| Field | Type | Inclusion | Description |
+| ----: | :--: | :--: |----------- |
+| `topic` | `String` | mandatory | The [PubSub `topic`](https://github.com/libp2p/specs/blob/master/pubsub/README.md#the-topic-descriptor) being published on |
+| `message` | [`WakuRelayMessage`](#WakuRelayMessage) | mandatory | The `message` being relayed |
+
+#### Response
+
+- **`Bool`** - `true` on success or an [error](https://www.jsonrpc.org/specification#error_object) on failure.
+
+### `post_waku_v2_relay_v1_subscriptions`
+
+The `post_waku_v2_relay_v1_subscriptions` method subscribes a node to an array of [PubSub `topics`](https://github.com/libp2p/specs/blob/master/pubsub/README.md#the-topic-descriptor).
+
+#### Parameters
+
+| Field | Type | Inclusion | Description |
+| ----: | :--: | :--: |----------- |
+| `topics` | `Array`[`String`] | mandatory | The [PubSub `topics`](https://github.com/libp2p/specs/blob/master/pubsub/README.md#the-topic-descriptor) being subscribed to |
+
+#### Response
+
+- **`Bool`** - `true` on success or an [error](https://www.jsonrpc.org/specification#error_object) on failure.
+
+### `delete_waku_v2_relay_v1_subscriptions`
+
+The `delete_waku_v2_relay_v1_subscriptions` method unsubscribes a node from an array of [PubSub `topics`](https://github.com/libp2p/specs/blob/master/pubsub/README.md#the-topic-descriptor).
+
+#### Parameters
+
+| Field | Type | Inclusion | Description |
+| ----: | :--: | :--: |----------- |
+| `topics` | `Array`[`String`] | mandatory | The [PubSub `topics`](https://github.com/libp2p/specs/blob/master/pubsub/README.md#the-topic-descriptor) being unsubscribed from |
+
+#### Response
+
+- **`Bool`** - `true` on success or an [error](https://www.jsonrpc.org/specification#error_object) on failure.
+
+### `get_waku_v2_relay_v1_messages_head`
+
+The `get_waku_v2_relay_v1_messages_head` method returns the latest message on a subscribed [PubSub `topic`](https://github.com/libp2p/specs/blob/master/pubsub/README.md#the-topic-descriptor). The server MUST respond with an [error](https://www.jsonrpc.org/specification#error_object) if no subscription exists for the polled `topic`. If no message has yet been received on the polled `topic`, the server MUST block until a message is received or a timeout is reached. This method can be used to long poll a `topic` for new messages.
```

Comment:
I wonder how useful this "latest message" will be, practically. For example, what if you call it and you have received 10 messages in meantime? In v1 it works like this:

> Retrieves messages that match a filter criteria and were received after the last time this function was called.

Which requires keeping some state. I wonder if something similar would make sense? perhaps with a `LIMIT N` query? Not sure.

This is probably an enhancement issue in such a case.

Reply:
True. The issue for me here is that I can't really see how polling can be done without keeping some state. Getting only the `head` would be a simple way of proving the concept and keeping state to a minimum.
The endpoint here is specified in such a way that it does not preclude adding `get_waku_v2_relay_v1_messages` which will retrieve messages matching a filter criteria and "were received after the last time this function was called". This can then be called with an optional `LIMIT N`.
However, I'm getting more and more convinced that the correct "libp2p"-way of doing this would be to register an event listener that pushes notifications to the client. This will require support for a websocket server. Seems like nimbus-eth2 is also deferring [event listening](https://github.com/status-im/nimbus-eth2/blob/master/beacon_chain/rpc/event_api.nim) on the RPC API for now.